runtime/compiler/runtime/JITServerAOTCache.cpp

```diff
@@ -745,6 +1143,325 @@ JITServerAOTCache::printStats(FILE *f) const
+      }
+      {
+      OMR::CriticalSection cs(_classLoaderMonitor);
+      header._numClassLoaderRecords = _classLoaderMap.size();
+      }
+
+   if (1 != fwrite(&header, sizeof(JITServerAOTCacheHeader), 1, f))
+      return false;
+
+   if (!writeAOTRecordList(f, _classLoaderHead, header._numClassLoaderRecords))
+      return false;
+   if (!writeAOTRecordList(f, _classHead, header._numClassRecords))
+      return false;
+   if (!writeAOTRecordList(f, _methodHead, header._numMethodRecords))
+      return false;
+   if (!writeAOTRecordList(f, _classChainHead, header._numClassChainRecords))
+      return false;
+   if (!writeAOTRecordList(f, _wellKnownClassesHead, header._numWellKnownClassesRecords))
+      return false;
+   if (!writeAOTRecordList(f, _aotHeaderHead, header._numAOTHeaderRecords))
+      return false;
+   if (!writeCachedMethodList(f, _cachedMethodHead, header._numCachedAOTMethods))
+      return false;
+
+   return true;
+   }
+
+// Tests whether or not the given AOT snapshot is compatible with the server.
+static bool
+isCompatibleSnapshotVersion(const JITServerAOTCacheVersion *version)
+   {
+   JITServerAOTCacheVersion currentVersion = {0};
+   getCurrentAOTCacheVersion(&currentVersion);
+
+   return (0 == strncmp(version->_eyeCatcher, currentVersion._eyeCatcher, sizeof(currentVersion._eyeCatcher))) &&
+          (version->_snapshotVersion == currentVersion._snapshotVersion) &&
+          (version->_jitserverVersion == currentVersion._jitserverVersion);
+   }
+
+// Read an AOT cache snapshot, returning NULL if the cache is ill-formed or
+// incompatible with the running server.
+JITServerAOTCache *
+JITServerAOTCache::readCache(FILE *f, const std::string &name)
+   {
+   JITServerAOTCacheHeader header = {0};
+   if (1 != fread(&header, sizeof(JITServerAOTCacheHeader), 1, f))
+      return NULL;
+
+   if (!isCompatibleSnapshotVersion(&header._version))
+      return NULL;
+
+   JITServerAOTCache *cache = new (TR::Compiler->persistentGlobalMemory()) JITServerAOTCache(name);
+   if (!cache)
+      throw std::bad_alloc();
+
+   if (!cache->readCache(f, &header))
+      {
+      cache->~JITServerAOTCache();
+      TR::Compiler->persistentGlobalMemory()->freePersistentMemory(cache);
+      return NULL;
+      }
+
+   return cache;
+   }
+
+bool
+JITServerAOTCache::readCache(FILE *f, JITServerAOTCacheHeader *header)
+   {
+   _classLoaderMap.reserve(header->_numClassLoaderRecords);
+   _classMap.reserve(header->_numClassRecords);
+   _methodMap.reserve(header->_numMethodRecords);
+   _classChainMap.reserve(header->_numClassChainRecords);
+   _wellKnownClassesMap.reserve(header->_numWellKnownClassesRecords);
+   _aotHeaderMap.reserve(header->_numAOTHeaderRecords);
+   _cachedMethodMap.reserve(header->_numCachedAOTMethods);
+
+   // ID 0 denotes an invalid record, so the next IDs and ID vectors
+   // must all have these sizes
+   _nextClassLoaderId = header->_numClassLoaderRecords + 1;
```

Comment:
While this is correct in the current implementation, if we ever allow removing stale methods and records from the cache (while taking a snapshot), this will no longer be true. I think we may as well deal with this now by storing the maximum IDs for each type in the snapshot header.